### ISSUE TITLE
Various issues with borders/drawing lines

### ISSUE THREAD

**Issue by alekseyt** [vurtun/nuklear#603]
I modified demo app (SDL2/opengl) in the following way: it creates 8 windows in donut shape named 11, 12, 13, ...,  33, then creates a single group in window 11 (top left). WM window is 300x300 pixels, each Nuklear window is 100x100 pixels, so mapping of windows to pixels is pretty straightforward, 3 windows in a row, each 100 pixels, 3 windows are covering entire window precisely, no floating point coords/sizes.

When anti-aliasing is "on" those bugs are not quite visible, so AA is need to be turned off too. Various paddings and spacings are set to zero too to exclude them from equations.

Complete modified example on this gist: https://gist.github.com/alekseyt/e6d710a8fd12e6ac6a8ebad523bd1811 . 

In a nutshell, important part of the example is:
```
nk_begin(ctx, "11", nk_rect(0, 0, 100, 100), window_flags);
nk_layout_row_dynamic(ctx, 100, 1);
nk_group_begin(ctx, "g11", NK_WINDOW_BORDER | NK_WINDOW_NO_SCROLLBAR);
nk_group_end(ctx);
nk_end(ctx);
```

And
```
nk_sdl_render(NK_ANTI_ALIASING_OFF);
```

This example produces the following picture (zoomed 2x no interpolation):

![screen1](https://user-images.githubusercontent.com/993454/34752436-04dda254-f5bb-11e7-85aa-ce0435b30eba.png)

Green and red text/lines are my annotations on top of rendered graphics.

You can spot several artifacts on it, namely:
1. Whitish lines between windows 12/13, 31/32/33 (doesn't reproduce with AA on)
2. No group border at left and top
3. Group border appears to be shifted left relatively to window 11

I've tried to debug this with mixed success. I was able to transform this picture into the following:

![screen2](https://user-images.githubusercontent.com/993454/34752520-77bda80a-f5bb-11e7-9470-47226f9e0509.png)

Not that while lines between windows disappeared, top group border is now visible, but bottom border disappeared, border also seems to be more aligned to window, but not quite.

Patch:
```
diff --git a/nuklear.h b/nuklear.h
index bd78ec5..ee8ffb8 100644
--- a/nuklear.h
+++ b/nuklear.h
@@ -285,6 +285,7 @@ extern "C" {
 #define NK_MIN(a,b) ((a) < (b) ? (a) : (b))
 #define NK_MAX(a,b) ((a) < (b) ? (b) : (a))
 #define NK_CLAMP(i,v,x) (NK_MAX(NK_MIN(v,x), i))
+#define NK_AA(x) (floor((x) + 0.5))
 
 #ifdef NK_INCLUDE_STANDARD_VARARGS
   #if defined(_MSC_VER) && (_MSC_VER >= 1600) /* VS 2010 and above */
@@ -9646,8 +9647,8 @@ nk_draw_list_stroke_line(struct nk_draw_list *list, struct nk_vec2 a,
         nk_draw_list_path_line_to(list, a);
         nk_draw_list_path_line_to(list, b);
     } else {
-        nk_draw_list_path_line_to(list, nk_vec2_sub(a,nk_vec2(0.5f,0.5f)));
-        nk_draw_list_path_line_to(list, nk_vec2_sub(b,nk_vec2(0.5f,0.5f)));
+        nk_draw_list_path_line_to(list, a);
+        nk_draw_list_path_line_to(list, b);
     }
     nk_draw_list_path_stroke(list,  col, NK_STROKE_OPEN, thickness);
 }
@@ -9663,7 +9664,7 @@ nk_draw_list_fill_rect(struct nk_draw_list *list, struct nk_rect rect,
         nk_draw_list_path_rect_to(list, nk_vec2(rect.x, rect.y),
             nk_vec2(rect.x + rect.w, rect.y + rect.h), rounding);
     } else {
-        nk_draw_list_path_rect_to(list, nk_vec2(rect.x-0.5f, rect.y-0.5f),
+        nk_draw_list_path_rect_to(list, nk_vec2(rect.x, rect.y),
             nk_vec2(rect.x + rect.w, rect.y + rect.h), rounding);
     } nk_draw_list_path_fill(list,  col);
 }
@@ -9678,7 +9679,7 @@ nk_draw_list_stroke_rect(struct nk_draw_list *list, struct nk_rect rect,
         nk_draw_list_path_rect_to(list, nk_vec2(rect.x, rect.y),
             nk_vec2(rect.x + rect.w, rect.y + rect.h), rounding);
     } else {
-        nk_draw_list_path_rect_to(list, nk_vec2(rect.x-0.5f, rect.y-0.5f),
+        nk_draw_list_path_rect_to(list, nk_vec2(rect.x, rect.y),
             nk_vec2(rect.x + rect.w, rect.y + rect.h), rounding);
     } nk_draw_list_path_stroke(list,  col, NK_STROKE_CLOSED, thickness);
 }
@@ -19035,6 +19036,8 @@ nk_panel_end(struct nk_context *ctx)
             layout->bounds.y + layout->bounds.h + layout->footer_height:
             window->bounds.y + window->bounds.h;
 
+        //nk_stroke_rect(out, window->bounds, 0, layout->border, border_color);
+
         /* draw border top */
         nk_stroke_line(out,window->bounds.x,window->bounds.y,
             window->bounds.x + window->bounds.w, window->bounds.y,
@@ -19045,13 +19048,13 @@ nk_panel_end(struct nk_context *ctx)
             window->bounds.x + window->bounds.w, padding_y, layout->border, border_color);
 
         /* draw left border */
-        nk_stroke_line(out, window->bounds.x + layout->border*0.5f,
-            window->bounds.y, window->bounds.x + layout->border*0.5f,
+        nk_stroke_line(out, window->bounds.x,
+            window->bounds.y, window->bounds.x,
             padding_y, layout->border, border_color);
 
         /* draw right border */
-        nk_stroke_line(out, window->bounds.x + window->bounds.w - layout->border*0.5f,
-            window->bounds.y, window->bounds.x + window->bounds.w - layout->border*0.5f,
+        nk_stroke_line(out, window->bounds.x + window->bounds.w,
+            window->bounds.y, window->bounds.x + window->bounds.w,
             padding_y, layout->border, border_color);
     }
```

Basically i've found couple of suspicious places in code:

1. In  nk_convert(): i'm not sure what's up with all the -0.5f stuff, i think it was supposed to somehow round values to integers before, judging by commit history, but i don't know what subtracting 0.5 is supposed to do now.

If i "fix" rounding to integers with NK_AA macro (in patch), it doesn't really help, it has same effect if i just remove subtraction from code, so i left macro in patch, but didn't use it. 

However removing subtraction fixes white lines between windows.

2. In nk_panel_end():
```
/* draw left border */
nk_stroke_line(out, window->bounds.x + layout->border*0.5f,
     window->bounds.y, window->bounds.x + layout->border*0.5f,
     padding_y, layout->border, border_color);
```

I believe this is incorrect. Border width is 1 pixel, shifting cursor half-border to the right will produce border in the middle of pixels. I think it should draw border at window->bounds.x (but it doesn't).

```
/* draw right border */
nk_stroke_line(out, window->bounds.x + window->bounds.w - layout->border*0.5f,
    window->bounds.y, window->bounds.x + window->bounds.w - layout->border*0.5f,
    padding_y, layout->border, border_color);
```

Same as above, i think it should draw border at window->bounds.x + window->bounds.w - layout->border. Geometrically speaking, so to say, but again, it doesn't.

3. Note that i have ``nk_stroke_rect(out, window->bounds, 0, layout->border, border_color);`` commented out in nk_panel_end(). If i replace panel border drawing using lines with this single call, it solves most border issues in my app (but not in the linked gist). I'm using renderer different from plain OpenGL, so i'm not sure what exactly is fixing the issue.

Any suggestions?

**Comment by alekseyt** [comment]
After fiddling with it a little bit more, i think the way to go is actually
```
nk_sdl_render(NK_ANTI_ALIASING_ON);
```
Because it disables some transformations and equivalent to manually removing -0.5f in code. However i still think there is a bug somewhere within NK_ANTI_ALIASING_OFF. It's also probably worth noting in doc that NK_ANTI_ALIASING_ON doesn't enables AA, but actually "delegates" AA to backend, or so i think.

I've also found that dynamic layout was subtracting one pixel from panel width at ``nuklear.h:20731``. I've also replaced line-by-line panel border drawing with single call to ``nk_stroke_rect()``, latter looks like a reasonable thing to do. 

Now demo from (slightly updated) gist looks like this:

![screenshot from 2018-01-10 16 27 44](https://user-images.githubusercontent.com/993454/34777692-cd3510f8-f623-11e7-9650-cac36be4abe3.png)

Gist update just added another group into window 12.

New and improved patch:
```
diff --git a/nuklear.h b/nuklear.h
index bd78ec5..4cdf70e 100644
--- a/nuklear.h
+++ b/nuklear.h
@@ -19035,24 +19035,7 @@ nk_panel_end(struct nk_context *ctx)
             layout->bounds.y + layout->bounds.h + layout->footer_height:
             window->bounds.y + window->bounds.h;
 
-        /* draw border top */
-        nk_stroke_line(out,window->bounds.x,window->bounds.y,
-            window->bounds.x + window->bounds.w, window->bounds.y,
-            layout->border, border_color);
-
-        /* draw bottom border */
-        nk_stroke_line(out, window->bounds.x, padding_y,
-            window->bounds.x + window->bounds.w, padding_y, layout->border, border_color);
-
-        /* draw left border */
-        nk_stroke_line(out, window->bounds.x + layout->border*0.5f,
-            window->bounds.y, window->bounds.x + layout->border*0.5f,
-            padding_y, layout->border, border_color);
-
-        /* draw right border */
-        nk_stroke_line(out, window->bounds.x + window->bounds.w - layout->border*0.5f,
-            window->bounds.y, window->bounds.x + window->bounds.w - layout->border*0.5f,
-            padding_y, layout->border, border_color);
+        nk_stroke_rect(out, window->bounds, 0, layout->border, border_color);
     }
 
     /* scaler */
@@ -20728,7 +20711,7 @@ nk_layout_widget_space(struct nk_rect *bounds, const struct nk_context *ctx,
     switch (layout->row.type) {
     case NK_LAYOUT_DYNAMIC_FIXED: {
         /* scaling fixed size widgets item width */
-        item_width = NK_MAX(1.0f,panel_space-1.0f) / (float)layout->row.columns;
+        item_width = NK_MAX(1.0f,panel_space) / (float)layout->row.columns;
         item_offset = (float)layout->row.index * item_width;
         item_spacing = (float)layout->row.index * spacing.x;
     } break;
```
I don't know about ``+-layout->border*0.5f``. I understand why border is need to be centered in case if border width is greater than 1, but ``bounds.x + bounds.w - border*0.5`` doesn't look right. If x=0, w=100 and border=1, then ``0+100-1*0.5=99.5``, then it's truncated to short which is 99, right? This math just looks wrong, but ``nk_stroke_rect()`` seems to work alright in my case.

I think this is definitely a place worth looking into. But patch above supposedly could at least workaround some of the mentioned issues.

**Comment by vurtun** [comment]
Hey @alekseyt,
thanks for all your work. I will take a closer look as soon as I can. I remember that border was an absolute nightmare and pain. Problem is not only the vertex buffer backend but also the more high-level backends like X11 or win32. Often if you remove something or change something using one backends as reference it will go wrong in another.

**Comment by alekseyt** [comment]
Yeah, no probs. I already applied patch to my copy of Nuklear, but obviously i don't want to maintain my own fork, hope patch helps. I originally stumbled on this issue when trying to make dock-like panel at the bottom of the screen, but my dock wasn't one huge panel but a bunch of windows arranged in supposedly seamless fashion. Unfortunately "seamless fashion" had some pixel-wide gaps in it and due to semi-transparent theme, background pixels where "shining" through this gap.

It appears that main suspect for that is 
```
-        item_width = NK_MAX(1.0f,panel_space-1.0f) / (float)layout->row.columns;
+        item_width = NK_MAX(1.0f,panel_space) / (float)layout->row.columns;
```
which was leaving exactly one pixel gap at the right side of widgets.

The border drawing code was just acting strange, for example, in my situation right border was wider than left border.

I'm still seeing some artifacts on a different rendered, not on the plain SDL/OpenGL and this time not related to borders though, so i will be looking into this too/again, eventually.

Apparently patch from previous comment didn't cover up case when window is minimized. Top-most part of the patch doesn't fix anything, just improves readability (in my opinion) and not really required. Important change is:
```
+        struct nk_rect border_rect = nk_rect(window->bounds.x, window->bounds.y,
+            window->bounds.w, padding_y - window->bounds.y);
+        nk_stroke_rect(out, border_rect, 0, layout->border, border_color);
```

Full patch:
```
diff --git a/nuklear.h b/nuklear.h
index bd78ec5..48cb71e 100644
--- a/nuklear.h
+++ b/nuklear.h
@@ -19029,30 +19029,15 @@ nk_panel_end(struct nk_context *ctx)
     if (layout->flags & NK_WINDOW_BORDER)
     {
         struct nk_color border_color = nk_panel_get_border_color(style, layout->type);
-        const float padding_y = (layout->flags & NK_WINDOW_MINIMIZED) ?
-            style->window.border + window->bounds.y + layout->header_height:
-            (layout->flags & NK_WINDOW_DYNAMIC)?
-            layout->bounds.y + layout->bounds.h + layout->footer_height:
-            window->bounds.y + window->bounds.h;
+        const float padding_y = (layout->flags & NK_WINDOW_MINIMIZED)
+            ? (style->window.border + window->bounds.y + layout->header_height)
+            : ((layout->flags & NK_WINDOW_DYNAMIC)
+                ? layout->bounds.y + layout->bounds.h + layout->footer_height
+                : window->bounds.y + window->bounds.h);
 
-        /* draw border top */
-        nk_stroke_line(out,window->bounds.x,window->bounds.y,
-            window->bounds.x + window->bounds.w, window->bounds.y,
-            layout->border, border_color);
-
-        /* draw bottom border */
-        nk_stroke_line(out, window->bounds.x, padding_y,
-            window->bounds.x + window->bounds.w, padding_y, layout->border, border_color);
-
-        /* draw left border */
-        nk_stroke_line(out, window->bounds.x + layout->border*0.5f,
-            window->bounds.y, window->bounds.x + layout->border*0.5f,
-            padding_y, layout->border, border_color);
-
-        /* draw right border */
-        nk_stroke_line(out, window->bounds.x + window->bounds.w - layout->border*0.5f,
-            window->bounds.y, window->bounds.x + window->bounds.w - layout->border*0.5f,
-            padding_y, layout->border, border_color);
+        struct nk_rect border_rect = nk_rect(window->bounds.x, window->bounds.y,
+            window->bounds.w, padding_y - window->bounds.y);
+        nk_stroke_rect(out, border_rect, 0, layout->border, border_color);
     }
 
     /* scaler */
@@ -20728,7 +20713,7 @@ nk_layout_widget_space(struct nk_rect *bounds, const struct nk_context *ctx,
     switch (layout->row.type) {
     case NK_LAYOUT_DYNAMIC_FIXED: {
         /* scaling fixed size widgets item width */
-        item_width = NK_MAX(1.0f,panel_space-1.0f) / (float)layout->row.columns;
+        item_width = NK_MAX(1.0f,panel_space) / (float)layout->row.columns;
         item_offset = (float)layout->row.index * item_width;
         item_spacing = (float)layout->row.index * spacing.x;
     } break;
```

Sorry for a lot of patches, i'm not really expert in Nuklear, i'm figuring this out as i go.

**Comment by vurtun** [comment]
I tested your code and both the vertex buffer output as well as X11. Your patch solved the vertex buffer backend. But it seems to me that X11 was the perpetrator in the past. It seems like X11 draws left and top at the beginning pixel and overdraws the right and bottom side. So the X11 backend needs to decrement the with and height by one.

After applying your patch for vertex buffer I got:

![sdl](https://user-images.githubusercontent.com/8057201/35481191-376cab94-041f-11e8-9112-802b14942c7d.png)

and for X11 after an additional decrement change of width and height I got:
![x11](https://user-images.githubusercontent.com/8057201/35481192-390be082-041f-11e8-9d3e-9390fe9c8ee7.png)

which seems correct now.
